packages/storage-plus/src/stack.rs

```diff
@@ -0,0 +1,652 @@
+    /// gets the length from storage, and otherwise sets it to 0
+    pub fn get_len(&self, storage: &dyn Storage) -> StdResult<u32> {
+        let mut may_len = self.length.lock().unwrap();
+        match *may_len {
+            Some(len) => Ok(len),
+            None => {
+                let len_key = [self.as_slice(), LEN_KEY].concat();
+                if let Some(len_vec) = storage.get(&len_key) {
+                    let len_bytes = len_vec
+                        .as_slice()
+                        .try_into()
+                        .map_err(|err| StdError::parse_err("u32", err))?;
+                    let len = u32::from_be_bytes(len_bytes);
+                    *may_len = Some(len);
+                    Ok(len)
+                } else {
+                    *may_len = Some(0);
+                    Ok(0)
+                }
+            }
+        }
+    }
+    /// checks if the collection has any elements
+    pub fn is_empty(&self, storage: &dyn Storage) -> StdResult<bool> {
+        Ok(self.get_len(storage)? == 0)
+    }
+    /// gets the element at pos if within bounds
+    pub fn get_at(&self, storage: &dyn Storage, pos: u32) -> StdResult<T> {
+        let len = self.get_len(storage)?;
+        if pos > len {
+            return Err(StdError::generic_err("Stack access out of bounds"));
+        }
+        self.get_at_unchecked(storage, pos)
+    }
+    /// tries to get the element at pos
+    fn get_at_unchecked(&self, storage: &dyn Storage, pos: u32) -> StdResult<T> {
+        let key = pos.to_be_bytes();
+        self.load_impl(storage, &key)
+    }
+
+    /// Set the length of the collection
+    fn set_len(&self, storage: &mut dyn Storage, len: u32) {
+        let len_key = [self.as_slice(), LEN_KEY].concat();
+        storage.set(&len_key, &len.to_be_bytes());
+
+        let mut may_len = self.length.lock().unwrap();
+        *may_len = Some(len);
+    }
+    /// Clear the collection
+    pub fn clear(&self, storage: &mut dyn Storage) {
+        self.set_len(storage, 0);
+    }
+    /// Replaces data at a position within bounds
+    pub fn set_at(&self, storage: &mut dyn Storage, pos: u32, item: &T) -> StdResult<()> {
+        let len = self.get_len(storage)?;
+        if pos >= len {
+            return Err(StdError::generic_err("Stack access out of bounds"));
+        }
+        self.set_at_unchecked(storage, pos, item)
+    }
+    /// Sets data at a given index
+    fn set_at_unchecked(&self, storage: &mut dyn Storage, pos: u32, item: &T) -> StdResult<()> {
+        self.save_impl(storage, &pos.to_be_bytes(), item)
+    }
+    /// Pushes an item to Stack
+    pub fn push(&self, storage: &mut dyn Storage, item: &T) -> StdResult<()> {
+        let len = self.get_len(storage)?;
+        self.set_at_unchecked(storage, len, item)?;
+        self.set_len(storage, len + 1);
+        Ok(())
+    }
+    /// Pops an item from Stack
+    pub fn pop(&self, storage: &mut dyn Storage) -> StdResult<T> {
```

Comment:
I think this should return `Option<T>` and nor error in case there is no element left